### ISSUE TITLE
Error occurs when inferring a scripted module with libtorch c++ on a cuda device 

### ISSUE THREAD

**Issue by Link2027** [pytorch/pytorch#48479]
## 🐛 Bug

When I was inferring a scripted transformer model with libtorch c++ **on a CPU**, everything was ok.
The scripted transformer can also generate the correct result for the first tensor input on **a cuda device**.
However, when inferring the second tensor input on **a cuda device**, I got the following error:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  default_program(22): error: more than one instance of overloaded function "fmod" matches the argument list:
            function "fmod(double, double)"
            function "fmod(float, float)"
            argument types are: (long long, long long)

default_program(24): error: more than one instance of overloaded function "fmod" matches the argument list:
            function "fmod(double, double)"
            function "fmod(float, float)"
            argument types are: (long long, long long)

2 errors detected in the compilation of "default_program".

nvrtc compilation failed:

#define NAN __int_as_float(0x7fffffff)
#define POS_INFINITY __int_as_float(0x7f800000)
#define NEG_INFINITY __int_as_float(0xff800000)


template<typename T>
__device__ T maximum(T a, T b) {
  return isnan(a) ? a : (a > b ? a : b);
}

template<typename T>
__device__ T minimum(T a, T b) {
  return isnan(a) ? a : (a < b ? a : b);
}

extern "C" __global__
void func_2(int v0, long long* t1, int v2, bool* aten_eq_flat, long long* aten_fmod_flat) {
{
  if (512 * blockIdx.x + threadIdx.x<2 ? 1 : 0) {
    long long v = __ldg(t1 + (512 * blockIdx.x + threadIdx.x) % 2);
    aten_fmod_flat[512 * blockIdx.x + threadIdx.x] = fmod(v, (long long)(v2));
    long long v_1 = __ldg(t1 + (512 * blockIdx.x + threadIdx.x) % 2);
    aten_eq_flat[512 * blockIdx.x + threadIdx.x] = (bool)((fmod(v_1, (long long)(v2)))==(long long)(v0) ? 1 : 0);
  }
}
}
```

## To Reproduce

libtorch c++ code:

```
  // load torchscript model
  torch::jit::script::Module module;
  module = torch::jit::load(argv[1]);
  module.to(at::kCUDA);

  // start evaluation
  long data[] = {4,   30, 1041, 1645, 244, 2};
  torch::Tensor sample = torch::from_blob(data, {1, 6}, torch::TensorOptions().dtype(torch::kLong));
  module.forward({sample.to(torch::kCUDA)});   // no error occurred
  module.forward({sample.to(torch::kCUDA)});   // crashed here

```

## Expected behavior

No error occurs.

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.0+cu110
 - OS (e.g., Linux):  Ubuntu 18.04.3 LTS (x86_64)
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): 
 - Python version: 3.6 (64-bit runtime)
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: GeForce RTX 2080 Ti
 - Any other relevant information:

## Additional context

The transformer was trained with fairseq on GPU.

**Comment by Link2027** [comment]
Hi, I managed to solve this issue by changing fairseq code:
indices_buf = indices_buf.fmod(vocab_size)
->
indices_buf = indices_buf.fmod(float(vocab_size))

**Comment by afpapqy** [comment]
The same problem! Have you solved it?

**Comment by kuynzereb** [comment]
It is a problem with `fmod` in 1.7.* versions. One can reproduce this error with the following minimal example:

```
import torch


def func(t_1: torch.Tensor, t_2: torch.Tensor):
    result = (t_1 + t_2) % 2

    return result


jit_func = torch.jit.script(func)

t_1 = torch.tensor([0], dtype=torch.long, device='cuda:0')
t_2 = torch.tensor([0], dtype=torch.long, device='cuda:0')

jit_func(t_1, t_2)
jit_func(t_1, t_2)
```

For example, I got the following output when run it with pytorch 1.7.1:
```
Traceback (most recent call last):
  File "sandbox/minimal_example.py", line 16, in <module>
    jit_func(t_1, t_2)
RuntimeError: default_program(23): error: more than one instance of overloaded function "fmod" matches the argument list:
            function "fmod(double, double)"
            function "fmod(float, float)"
            argument types are: (long long, int)

1 error detected in the compilation of "default_program".

nvrtc compilation failed: 

#define NAN __int_as_float(0x7fffffff)
#define POS_INFINITY __int_as_float(0x7f800000)
#define NEG_INFINITY __int_as_float(0xff800000)


template<typename T>
__device__ T maximum(T a, T b) {
  return isnan(a) ? a : (a > b ? a : b);
}

template<typename T>
__device__ T minimum(T a, T b) {
  return isnan(a) ? a : (a < b ? a : b);
}

extern "C" __global__
void func_1(long long* t0, long long* t1, long long* aten_remainder_flat) {
{
  if (512 * blockIdx.x + threadIdx.x<1 ? 1 : 0) {
    long long v = __ldg(t1 + 0);
    long long v_1 = __ldg(t0 + 0);
    aten_remainder_flat[512 * blockIdx.x + threadIdx.x] = fmod((fmod(v_1 + v, 2)) + 2, 2);
  }
}
}

```

But it works fine with nightly build. So I guess this problem will be fixed in 1.8.0.

**Comment by afpapqy** [comment]
thanks a lot






------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Vadim Bereznyuk"<notifications@github.com&gt;; 
发送时间: 2021年1月23日(星期六) 凌晨0:21
收件人: "pytorch/pytorch"<pytorch@noreply.github.com&gt;; 
抄送: "1504付攀"<1090146748@qq.com&gt;; "Comment"<comment@noreply.github.com&gt;; 
主题: Re: [pytorch/pytorch] Error occurs when inferring a scripted module with libtorch c++ on a cuda device  (#48479)





 
It is a problem with fmod in 1.7.* versions. One can reproduce this error with the following minimal example:
 import torch def func(t_1: torch.Tensor, t_2: torch.Tensor):     result = (t_1 + t_2) % 2     return result jit_func = torch.jit.script(func) t_1 = torch.tensor([0], dtype=torch.long, device='cuda:0') t_2 = torch.tensor([0], dtype=torch.long, device='cuda:0') jit_func(t_1, t_2) jit_func(t_1, t_2)  
For example, I got the following output when run it with pytorch 1.7.1:
 Traceback (most recent call last):   File "sandbox/minimal_example.py", line 16, in <module&gt;     jit_func(t_1, t_2) RuntimeError: default_program(23): error: more than one instance of overloaded function "fmod" matches the argument list:             function "fmod(double, double)"             function "fmod(float, float)"             argument types are: (long long, int) 1 error detected in the compilation of "default_program". nvrtc compilation failed:  #define NAN __int_as_float(0x7fffffff) #define POS_INFINITY __int_as_float(0x7f800000) #define NEG_INFINITY __int_as_float(0xff800000) template<typename T&gt; __device__ T maximum(T a, T b) {   return isnan(a) ? a : (a &gt; b ? a : b); } template<typename T&gt; __device__ T minimum(T a, T b) {   return isnan(a) ? a : (a < b ? a : b); } extern "C" __global__ void func_1(long long* t0, long long* t1, long long* aten_remainder_flat) { {   if (512 * blockIdx.x + threadIdx.x<1 ? 1 : 0) {     long long v = __ldg(t1 + 0);     long long v_1 = __ldg(t0 + 0);     aten_remainder_flat[512 * blockIdx.x + threadIdx.x] = fmod((fmod(v_1 + v, 2)) + 2, 2);   } } }  
But it works fine with nightly build. So I guess this problem will be fixed in 1.8.0.
